builtin/credential/ldap/backend.go

```diff
@@ -227,213 +194,6 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
 	return policies, ldapResponse, allGroups, nil
 }
 
-/*
- * Parses a distinguished name and returns the CN portion.
- * Given a non-conforming string (such as an already-extracted CN),
- * it will be returned as-is.
- */
-func (b *backend) getCN(dn string) string {
-	parsedDN, err := ldap.ParseDN(dn)
-	if err != nil || len(parsedDN.RDNs) == 0 {
-		// It was already a CN, return as-is
-		return dn
-	}
-
-	for _, rdn := range parsedDN.RDNs {
-		for _, rdnAttr := range rdn.Attributes {
-			if rdnAttr.Type == "CN" {
-				return rdnAttr.Value
-			}
-		}
-	}
-
-	// Default, return self
-	return dn
-}
-
-/*
- * Discover and return the bind string for the user attempting to authenticate.
- * This is handled in one of several ways:
- *
- * 1. If DiscoverDN is set, the user object will be searched for using userdn (base search path)
- *    and userattr (the attribute that maps to the provided username).
- *    The bind will either be anonymous or use binddn and bindpassword if they were provided.
- * 2. If upndomain is set, the user dn is constructed as 'username@upndomain'. See https://msdn.microsoft.com/en-us/library/cc223499.aspx
- *
- */
-func (b *backend) getUserBindDN(cfg *ConfigEntry, c *ldap.Conn, username string) (string, error) {
-	bindDN := ""
-	if cfg.DiscoverDN || (cfg.BindDN != "" && cfg.BindPassword != "") {
-		var err error
-		if cfg.BindPassword != "" {
-			err = c.Bind(cfg.BindDN, cfg.BindPassword)
-		} else {
-			err = c.UnauthenticatedBind(cfg.BindDN)
-		}
-		if err != nil {
-			return bindDN, errwrap.Wrapf("LDAP bind (service) failed: {{err}}", err)
-		}
-
-		filter := fmt.Sprintf("(%s=%s)", cfg.UserAttr, ldap.EscapeFilter(username))
-		if b.Logger().IsDebug() {
-			b.Logger().Debug("discovering user", "userdn", cfg.UserDN, "filter", filter)
-		}
-		result, err := c.Search(&ldap.SearchRequest{
-			BaseDN:    cfg.UserDN,
-			Scope:     2, // subtree
-			Filter:    filter,
-			SizeLimit: math.MaxInt32,
-		})
-		if err != nil {
-			return bindDN, errwrap.Wrapf("LDAP search for binddn failed: {{err}}", err)
-		}
-		if len(result.Entries) != 1 {
-			return bindDN, fmt.Errorf("LDAP search for binddn 0 or not unique")
-		}
-		bindDN = result.Entries[0].DN
-	} else {
-		if cfg.UPNDomain != "" {
-			bindDN = fmt.Sprintf("%s@%s", EscapeLDAPValue(username), cfg.UPNDomain)
-		} else {
-			bindDN = fmt.Sprintf("%s=%s,%s", cfg.UserAttr, EscapeLDAPValue(username), cfg.UserDN)
-		}
-	}
-
-	return bindDN, nil
-}
-
-/*
- * Returns the DN of the object representing the authenticated user.
- */
-func (b *backend) getUserDN(cfg *ConfigEntry, c *ldap.Conn, bindDN string) (string, error) {
```

Comment:
This was likewise moved to the client without any internal change because it makes external calls. Exported as well to be usable outside the package.